FFmpegInterop/UncompressedSampleProvider.cpp

```diff
@@ -19,105 +19,65 @@
+		if (threadCount > 0)
 		{
-			// The decoder should have been drained and always ready to access input
-			_ASSERT(FALSE);
-			hr = E_UNEXPECTED;
+			m_codecContext->thread_count = threadCount;
+			m_codecContext->thread_type = FF_THREAD_FRAME | FF_THREAD_SLICE;
 		}
-		else if (sendPacketResult < 0)
-		{
-			// We failed to send the packet
-			hr = E_FAIL;
-			DebugMessage(L"Decoder failed on the sample\n");
-		}
-	}
-	if (SUCCEEDED(hr))
-	{
-		AVFrame *pFrame = av_frame_alloc();
-		// Try to get a frame from the decoder.
-		decodeFrame = avcodec_receive_frame(m_pAvCodecCtx, pFrame);
 
-		// The decoder is empty, send a packet to it.
-		if (decodeFrame == AVERROR(EAGAIN))
-		{
-			// The decoder doesn't have enough data to produce a frame,
-			// return S_FALSE to indicate a partial frame
-			hr = S_FALSE;
-			av_frame_unref(pFrame);
-			av_frame_free(&pFrame);
-		}
-		else if (decodeFrame < 0)
-		{
-			hr = E_FAIL;
-			av_frame_unref(pFrame);
-			av_frame_free(&pFrame);
-			DebugMessage(L"Failed to get a frame from the decoder\n");
-		}
-		else
-		{
-			m_pAvFrame = pFrame;
-		}
+		THROW_HR_IF_FFMPEG_FAILED(avcodec_open2(m_codecContext.get(), codec, nullptr));
 	}
 
-	return hr;
-}
+	void UncompressedSampleProvider::Flush() noexcept
+	{
+		SampleProvider::Flush();
 
-HRESULT UncompressedSampleProvider::DecodeAVPacket(DataWriter^ dataWriter, AVPacket* avPacket, int64_t& framePts, int64_t& frameDuration)
-{
-	HRESULT hr = S_OK;
-	bool fGotFrame  = false;
-	AVPacket *pPacket = avPacket;
+		avcodec_flush_buffers(m_codecContext.get());
+	}
 
-	while (SUCCEEDED(hr))
+	AVFrame_ptr UncompressedSampleProvider::GetFrame()
 	{
-		hr = GetFrameFromFFmpegDecoder(pPacket);
-		pPacket = nullptr;
-		if (SUCCEEDED(hr))
+		// Allocate a frame
+		AVFrame_ptr frame{ av_frame_alloc() };
+		THROW_IF_NULL_ALLOC(frame);
+
+		while (true)
 		{
-			if (hr == S_FALSE)
-			{
-				// If the decoder didn't give an initial frame we still need
-				// to feed it more frames. Keep S_FALSE as the result
-				if (fGotFrame)
-				{
-					hr = S_OK;
-				}
-				break;
-			}
-			// Update the timestamp if the packet has one
-			else if (m_pAvFrame->pts != AV_NOPTS_VALUE)
+			// Send a packet to the decoder and see if it can produce a frame
+			AVPacket_ptr packet{ GetPacket() };
+
+			THROW_HR_IF_FFMPEG_FAILED(avcodec_send_packet(m_codecContext.get(), packet.get()));
+
+			int decodeResult{ avcodec_receive_frame(m_codecContext.get(), frame.get()) };
+			if (decodeResult == AVERROR(EAGAIN))
 			{
-				framePts = m_pAvFrame->pts;
-				frameDuration = m_pAvFrame->pkt_duration;
+				// The decoder needs more data to produce a frame
+				TraceLoggingWrite(g_FFmpegInteropProvider, "DecoderNeedsMoreInput", TraceLoggingLevel(TRACE_LEVEL_VERBOSE), TraceLoggingPointer(this, "this"),
+					TraceLoggingValue(m_stream->index, "StreamId"));
+				continue;
 			}
-			fGotFrame = true;
+			THROW_HR_IF_FFMPEG_FAILED(decodeResult);
```

Comment:
>THROW_HR_IF_FFMPEG_FAILED(decodeResult); [](start = 3, length = 40)

FFmpegInterop previously had logic to allow a certain number of decode failures before failing out. If we don't think that it's worth maintaining that behavior, we should probably document this change for consumers (just in the commit message would be ok for now). I also noticed another behavior change in that we now fail playback when failing to get a sample for a stream, whereas previously the stream was disabled and playback continued.